### PR TITLE
feat: Add covfie magnetic field plugin

### PR DESCRIPTION
**Pull request by stephenswat**
This commit adds a new Acts plugin that adds support for magnetic fields implemented using the covfie library.

This commit is based on #3117.

Closes #3117.

Depends on #3478.

Virtually all credit goes to @fredevb.

### CODE REVIEW

**Comment by paulgessinger** [comment]
Can you update the branch? I'll look at this today

**Comment by stephenswat** [comment]
This is now ready for review.

File: Plugins/Covfie/src/FieldConversion.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+  }
+  return magneticField.getFieldUnchecked(position);
+}
+
+/// @brief Get the value of the field at a specific position of a general magnetic field.
+/// @param magneticField the interpolated magnetic field.
+/// @param cache the magnetic field cache.
+/// @param position the position of the field to look up.
+/// @return the field value at the given position.
+Acts::Vector3 getFieldEdgeInclusive(
```
Comment (paulgessinger) [inline Plugins/Covfie/src/FieldConversion.cpp:46]: I don't fully understand what this function is for.

File: Plugins/Covfie/src/FieldConversion.cpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+/// @brief Uses std::nextafter to generates a clamp backend
+/// configuration where arguments min and max hold floating point values.
+/// @param min (min_x, min_y, min_z)
+/// @param max (max_x, max_y, max_z)
+/// @return The clamp field configuration.
+template <typename backend_t>
+typename backend_t::configuration_t clampConfigurationFloat(
+    const std::vector<Acts::ActsScalar>& min,
+    const std::vector<Acts::ActsScalar>& max) {
+  return {{std::nextafter(static_cast<float>(min[0]),
```
Comment (paulgessinger) [inline Plugins/Covfie/src/FieldConversion.cpp:132]: Size check or a fixed sized span maybe?

File: Examples/Python/src/MagneticField.cpp [hunk 3]
```diff
@@ -36,12 +36,26 @@ using namespace pybind11::literals;
 
 namespace Acts::Python {
 
+/// @brief Get the field value without needing a cache.
+/// @note Does not return an Acts::Result, so will throw an error if the lookup fails.
+Acts::Vector3 getField(Acts::MagneticFieldProvider& self,
```
[the remark this answers is not in the document]
Comment (stephenswat) [reply Examples/Python/src/MagneticField.cpp:41]: I've aimed to solve this more generally in #3554.

File: Plugins/Covfie/src/FieldConversion.cpp (hunk 1 above)
Comment (stephenswat) [reply Plugins/Covfie/src/FieldConversion.cpp:46]: Removed.

**Comment by stephenswat** [comment]
@paulgessinger I have incorporated your feedback and rebased.

File: Plugins/Covfie/src/FieldConversion.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+#include "Acts/Utilities/AxisFwd.hpp"
+#include "Acts/Utilities/Grid.hpp"
+
+#include <cmath>
+#include <stdexcept>
+#include <type_traits>
+
+namespace Acts::CovfiePlugin {
+
+/// @brief Creates a strided covfie field that stores the values of the magnetic field in the volume given by min and max using a fixed sample spacing (determined by nPoints).
```
Comment (paulgessinger) [inline Plugins/Covfie/src/FieldConversion.cpp:21]: This didn't auto-format because of the `@`.

File: Plugins/Covfie/src/FieldConversion.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        p[1] = static_cast<float>(rv[1]);
+        p[2] = static_cast<float>(rv[2]);
+      }
+    }
+  }
+
+  return field;
+}
+
+/// @brief Generate the affine covfie configuration (scaling and rotation) given the size of the field (min and max)
```
Comment (paulgessinger) [inline Plugins/Covfie/src/FieldConversion.cpp:81]: Same here.

File: Plugins/Covfie/src/FieldConversion.cpp
Comment (paulgessinger) [inline Plugins/Covfie/src/FieldConversion.cpp:1]: Generally, you're free to comment functions with internal linkage in `cpp` files of course, but just for the record, we don't require it.

**Review by paulgessinger — COMMENTED** [review]
I'm not completely sure I understand how you get around the out-of-bounds field access with the interpolated fields now.

**Comment by stephenswat** [comment]
> I'm not completely sure I understand how you get around the out-of-bounds field access with the interpolated fields now.

The core problem is that ACTS won't allow me to access the points at the edge of a B-field. In a one-dimensional example, if I define a data series as $[2, 3, 2, 1]$, I can ask ACTS for the position at $x = 0$ and get $f(0) = 2$; I can also ask for e.g. $f(1.5) = 2.5$. But I cannot ask for $f(3)$ (which would be $1$) because of this line: https://github.com/acts-project/acts/blob/main/Core/include/Acts/MagneticField/InterpolatedBFieldMap.hpp#L256. Because $3$ is the boundary and the convention is that the range is half-open, this returns an error. My understanding is that this is by design.

The way I have solved (c.q. hacked) this is by asking for $f(0)$, $f(1)$, $f(2)$ as normal and then - when I get to $x = 3$, asking for $f(x')$ where $x'$ is the closest representable number that is smaller than $3$; so I end up asking for $2.999999999$ or something like that.

Not super clean (actually quite disgusting) but it works. I investigated accessing the underlying grid directly, but this is not generally supported by ACTS B-field classes, unfortunately.

**Comment by stephenswat** [comment]
I've resolved the comment issues; I've left them in even if they are not necessary because, well, why not. :smile:

**Comment by paulgessinger** [comment]
Ok, looks good. There's a few CI failures that need fixing, otherwise it's good to go I think.

**Comment by paulgessinger** [comment]
There's just one remaining clang tidy item about inconsistent argument names between declaration and definition.

**Review by andiwand — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [809b378e51249733567ae24bffe6cdffa9058503]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [report_pull_request](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882154)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882153)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882152)||
|:green_circle: | [test_ActsDumpGeometryIdentifiers](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882151)||
|:green_circle: | [test_ActsCheckObjectCountsCachedWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882150)||
|:green_circle: | [test_ActsCheckObjectCountsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882149)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882148)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882147)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882146)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882145)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882144)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882143)||
|:green_circle: | [test_ActsWorkflowCached](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882142)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882141)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882140)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882139)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882138)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882137)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882136)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882135)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882134)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882133)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882132)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882131)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882130)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882129)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882128)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882127)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882126)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882125)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882124)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882123)||
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/42882122)||

**Pull request merged**